flink-runtime/src/main/java/org/apache/flink/runtime/taskexecutor/partition/JobAwareShuffleEnvironment.java

```diff
@@ -0,0 +1,87 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.flink.runtime.taskexecutor.partition;
+
+import org.apache.flink.api.common.JobID;
+import org.apache.flink.metrics.MetricGroup;
+import org.apache.flink.runtime.deployment.ResultPartitionDeploymentDescriptor;
+import org.apache.flink.runtime.executiongraph.ExecutionAttemptID;
+import org.apache.flink.runtime.io.network.api.writer.ResultPartitionWriter;
+import org.apache.flink.runtime.io.network.partition.ResultPartitionID;
+import org.apache.flink.runtime.io.network.partition.consumer.InputGate;
+import org.apache.flink.runtime.shuffle.ShuffleEnvironment;
+
+import java.util.Collection;
+import java.util.function.Consumer;
+
+/**
+ * Job-aware variation of the {@link ShuffleEnvironment} interface.
+ *
+ * @see ShuffleEnvironment
+ */
+public interface JobAwareShuffleEnvironment<P extends ResultPartitionWriter, G extends InputGate> extends ShuffleEnvironment<P, G> {
+
+	/**
+	 * Factory method for the task's {@link ResultPartitionWriter}s.
+	 *
+	 * @param jobId the job id, used for tracking which job the created partitions belongs to
+	 * @param taskName the task name, used for logs
+	 * @param executionAttemptID execution attempt id of the task
+	 * @param resultPartitionDeploymentDescriptors descriptors of the partition, produced by the task
+	 * @param outputGroup shuffle specific group for output metrics
+	 * @param buffersGroup shuffle specific group for buffer metrics
+	 * @return collection of the task's {@link ResultPartitionWriter}s
+	 */
+	Collection<P> createResultPartitionWriters(
+		JobID jobId,
+		String taskName,
+		ExecutionAttemptID executionAttemptID,
+		Collection<ResultPartitionDeploymentDescriptor> resultPartitionDeploymentDescriptors,
+		MetricGroup outputGroup,
+		MetricGroup buffersGroup);
+
+	/**
+	 * Releases local resources occupied with the given partitions.
+	 */
+	void releaseFinishedPartitions(JobID jobId, Collection<ResultPartitionID> partitionIds);
```

Comment:
I'm wondering what effort it would be to keep track of the `ResultPartitionID -> JobID` mapping inside of the `JobAwareShuffleEnvironmentImpl`? If we would do this, then we would not have to introduce this method and we could have kept `TaskExecutorGateway#releasePartitions`. Now, the need for the `JobID` when releasing partitions needs to be looped through multiple components.

Reply:
Oh we could certainly do that, the introduction of the JobID parameter was never out of necessity, but purely so we _don't_ have to maintain an extra mapping. The JobID argument has the nice side-effect that you don't have to check the JobID for every single partition.